### ISSUE TITLE
gtk2reactor not servicing gui events when using producer/consumer or recursive callbacks

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#4489]
|<img alt="gabriele.lanaro's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">| gabriele.lanaro reported|
|-|-|
|Trac ID|trac#4489|
|Type|defect|
|Created|2010-06-05 07:00:25Z|
|Branch|https://github.com/twisted/twisted/tree/gtk2-io-deprioritize-4489|

The problem appeared in two test-cases, when producing a "series" (one following another) callbacks the problem appear.

The first test case I'm recursively adding callbacks to a deferred (after a pb call).

Test 1:
untar twistsample1.tar.gz
```
$ python runner.py
click on the button
```
minimize/unminimize window to force a widget redraw, the window should be blank.
you can click another time on the button, the event is catched but no "button animation" is performed

Test2:
launch twisted_block.py
You should have symilar symptoms.

Attachments:

* [twistsample.tar.gz](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/6bc/6bc4806f4c2a7973a4f7a657fe92d9f21ea97f9c/a57a6a56f9acaa583e4ee9cc7ab64247bf7ef32f.gz) (942 bytes) - added by gabriele.lanaro on 2010-06-05 07:01:25Z - First test case
* [twisted_block.py](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/6bc/6bc4806f4c2a7973a4f7a657fe92d9f21ea97f9c/616ed2ac11adcad06a6e8407b7455b3f1c857f73.py) (817 bytes) - added by gabriele.lanaro on 2010-06-05 07:01:56Z - second test case
* [gtk2unresponsive.py](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/6bc/6bc4806f4c2a7973a4f7a657fe92d9f21ea97f9c/3fe49bc4569392d5318bff2eefbd47f84446a97a.py) (722 bytes) - added by exarkun on 2010-06-07 20:18:56Z - 

<details><summary>Searchable metadata</summary>

```
trac-id__4489 4489
type__defect defect
reporter__gabriele_lanaro gabriele.lanaro
priority__normal normal
milestone__ 
branch__branches_gtk2_io_deprioritize_4489 branches/gtk2-io-deprioritize-4489
branch_author__exarkun exarkun
status__closed closed
resolution__fixed fixed
component__core core
keywords__ 
time__1275721225000000 1275721225000000
changetime__1332110142000000 1332110142000000
version__None None
owner__ 

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_1"></a>|@exarkun commented|
|-|-|

I can reproduce this on Ubuntu 9.10.  I can also reproduce it with a pure pygtk app (attached).  :/

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_2"></a>|@exarkun commented|
|-|-|

It's a problem on Ubuntu 10.04 as well.  I filed an upstream bug report.  We'll see what they say.  https://bugzilla.gnome.org/show_bug.cgi?id=620901

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_3"></a>|@exarkun commented|
|-|-|

(In [29243]) Branching to 'gtk2-io-deprioritize-4489'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_4"></a>|@exarkun commented|
|-|-|

(In [29244]) The minimal fix required

refs [#4489](https://github.com/twisted/twisted/issues/4489)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_5"></a>|@exarkun commented|
|-|-|

So the change I checked in to the branch makes gtk2reactor set an explicit priority for the I/O events we ask gobject to monitor for us.  The level selected is one that is lower priority than servicing GUI events (like expose).  This means rendering the GUI will be given priority over servicing network events.  I think this makes sense for GUI apps.

It's possible this makes sense as an application-configurable parameter of some sort, though.

I ran into problems testing this which may stem from [#4481](https://github.com/twisted/twisted/issues/4481) so I'm going back to that to try to finish it off before proposing this to be merged.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_7"></a>|@itamarst commented|
|-|-|

(In [33928]) Lower priority of I/O and timeout callbacks.

Refs [#5067](https://github.com/twisted/twisted/issues/5067) and [#4489](https://github.com/twisted/twisted/issues/4489)

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_8"></a>|@itamarst set status to `closed`|
|-|-|

(In [33931]) Merge gtk-ui-priority-5067: Lower priority for Twisted callbacks in gtk/glib.

Author: thomasvs, exarkun
Review: itamar
Fixes: [#4489](https://github.com/twisted/twisted/issues/4489), [#5067](https://github.com/twisted/twisted/issues/5067)

Lower the priority of Twisted events in the glib/gtk reactor so that the UI will always refresh.
